### DISCUSSION TITLE
sum of double not deterministic

### DISCUSSION THREAD

**Discussion by rdenham** [duckdb/duckdb#12693] [Q&A]
I've run into a small issue when I use `sum(..)` from a large parquet file. The field is double, and the value of the sum varies very slightly from call to call. The actual value doesn't worry me, it is accurate enough, it's just that it isn't deterministic, and I was hoping for a repeatable workflow. Here is an example:

```sql
SELECT 
  4877835.5 - SUM(areaha) AS adiff 
FROM 
  read_parquet('e2021_woodytab.parquet') 
WHERE 
  subbioregion IN (
    'Georgina Limestone', 
    'Southwestern Downs', 
    'Kynuna Plateau',
    'Southern Wooded Downs', 
    'Central Downs', 
    'Flinders', 
    'Barkly Tableland'
  );
```

The parquet file is available at https://www.data.qld.gov.au/dataset/6f0312c5-823b-4552-adb0-8aab48263c13/resource/31c6db4e-7e29-4015-98b0-249eb2a7fcc1/download/e2021_woodytab.parquet

If I run this a couple of times I get values like

```
┌────────────────────────┐
│         adiff          │
│         double         │
├────────────────────────┤
│ -8.381903171539307e-09 │
└────────────────────────┘

┌───────────────────────┐
│         adiff         │
│        double         │
├───────────────────────┤
│ 9.313225746154785e-09 │
└───────────────────────┘
```

I read the bit about [Floating point types](https://duckdb.org/docs/sql/data_types/numeric#floating-point-types), and I get that there are pitfalls with floating point arithmetic. Those pitfalls didn't seem that relevant to my case, but I was surprised at the lack of deterministic results.

Incidentally, the issue came up because in one calculation, the value was very close to 14.5, and for reporting we rounded that. Sometimes it came out as 14, and sometimes 15.  Which does sort of suggest that decimal would be more appropriate, but the data currently is stored  as  double.

If I cast to decimal then I don't have this issue. 

I was wondering if someone could tell me why sum isn't deterministic, and what people suggest I do to ensure repeatability.

This test was done on 

v0.10.1 4a89d97db8.

**Comment by Tishj** [comment]
I have to admit I didn't read the description very carefully, but it sounds like this is expected, and you should use something like `kahan_sum` instead?
<https://duckdb.org/docs/sql/aggregates#fsumarg>

If you want determinism you should probably run it with threads=1, the inconsistencies likely originate from floating point inaccuracies from using multiple threads.
I don't know if kahan_sum also needs threads=1

**Reply by rdenham** [reply]
Thanksyou @Tishj 

I tested kahan_sum (fsum), it is deterministic without setting threads=1. Standard sum with threads=1 is also deterministic. And as pointed out by others (@tonne1) my use case does suggeest that fixed point arithmetic would be more appropriate.

**Comment by tonne1** [comment]
Your use case is exactly the type of scenario where you would use fixed point arithmetic, since that guarantees the precision you request no matter how the result gets calculated.

**Reply by carlopi** [reply]
Agree, I would go with do the conversion explicitly on your side, where you set what's the precision relevant to you.

(I set an absurd one, given most areas will 1e-23 seems a bit too much to be useful)
```sql
SELECT 
    SUM(areaha::DECIMAL(38,23)) AS adiff 
  FROM 
    read_parquet('e2021_woodytab.parquet') 
  WHERE 
    subbioregion IN (
      'Georgina Limestone', 
      'Southwestern Downs', 
      'Kynuna Plateau',
      'Southern Wooded Downs', 
      'Central Downs', 
      'Flinders', 
      'Barkly Tableland'
    );
┌─────────────────────────────────┐
│              adiff              │
│         decimal(38,23)          │
├─────────────────────────────────┤
│ 4877835.49999999969641464070144 │
└─────────────────────────────────┘
```

**Comment by rdenham** [comment]
Thank you @carlopi, @tonne1, @Tishj

A quick summary:

* with threads > 1, sum of a double may be inconsistent (not a bug, expected behaviour)
* with threads = 1, sum will be consistent, but slower
* with casting to appropriate decimal sum will be consistent
* using [fsum or aliases sumKahan, kahan_sum](https://duckdb.org/docs/sql/aggregates#fsumarg) result will be consistent, and presumably accurate. Probably slightly slower than sum but not noticeable in my example.
